src/libstore/daemon.cc

```diff
@@ -944,16 +944,21 @@ void processConnection(
         throw Error("the Nix client version is too old");
 
     auto tunnelLogger = new TunnelLogger(to, clientVersion);
-    auto prevLogger = nix::logger;
+    auto prevLogger = nix::logger.get();
```

Comment:
You could std::move it and construct the TunnelLogger with `std::make_unique` instead.

Reply:
Here was the problem with this code:
If I `std::move` it to prevLogger, then `nix::logger` is invalid. It is only set to the _tunnelLogger_ if its `!recursive`; so it has to remain correct otherwise.

I also can't simply assign `nix::logger` to the _tunnelLogger_ as that will destroy the previousLogger which is called in the `Finally` block later.

It's a very tricky piece of code to get right with _unique_ptr_ -- perhaps `nix::logger` should be _shared_ptr_ instead or I can get make the _prevLogger_ a **const**